Sources/PackageGraph/DependencyResolver.swift

```diff
@@ -677,44 +747,57 @@ public class DependencyResolver<
         subjectTo allConstraints: ConstraintSet,
         excluding allExclusions: [Identifier: Set<Version>]
     ) -> AnySequence<AssignmentSet> {
-        func validVersions(_ container: Container) -> AnyIterator<Version> {
-            let constraints = allConstraints[container.identifier]
+        func validVersions(_ container: Container, in versionSet: VersionSetSpecifier) -> AnyIterator<Version> {
             let exclusions = allExclusions[container.identifier] ?? Set()
             var it = container.versions.reversed().makeIterator()
             return AnyIterator{ () -> Version? in
                     while let version = it.next() {
-                        if constraints.contains(version) && !exclusions.contains(version) {
+                        if versionSet.contains(version) && !exclusions.contains(version) {
                             return version
                     }
                 }
                 return nil
             }
         }
 
-        // Attempt to select each valid version in the preferred order.
+        // Helper method to abstract passing common parameters to merge().
         //
         // FIXME: We must detect recursion here.
-        return AnySequence(validVersions(container).lazy.flatMap{ version -> AnySequence<AssignmentSet> in
+        func merge(constraints: [Constraint], into assignment: AssignmentSet) -> AnySequence<AssignmentSet> {
+            return AnySequence(self.merge(
+                constraints: constraints,
+                into: assignment, subjectTo: allConstraints, excluding: allExclusions).lazy.map{ result in
+                // We found a complete valid assignment.
+                assert(result.checkIfValidAndComplete())
+                return result
+            })
+        }
+
+        switch allConstraints[container.identifier] {
+        case .unversioned(let constraints):
+            var assignment = AssignmentSet()
+            assignment[container] = .unversioned
```

Comment:
These two lines can also be factored into the helper function (with a parameter on the binding to set up)